src/Symfony/Component/Console/DependencyInjection/AddConsoleCommandPass.php

```diff
@@ -41,18 +41,21 @@ public function process(ContainerBuilder $container): void
             $definition->addTag('container.no_preload');
             $class = $container->getParameterBag()->resolveValue($definition->getClass());
 
-            if (isset($tags[0]['command'])) {
-                $aliases = $tags[0]['command'];
-            } else {
-                if (!$r = $container->getReflectionClass($class)) {
-                    throw new InvalidArgumentException(\sprintf('Class "%s" used for service "%s" cannot be found.', $class, $id));
-                }
-                if (!$r->isSubclassOf(Command::class)) {
-                    throw new InvalidArgumentException(\sprintf('The service "%s" tagged "%s" must be a subclass of "%s".', $id, 'console.command', Command::class));
+            if (!$r = $container->getReflectionClass($class)) {
+                throw new InvalidArgumentException(\sprintf('Class "%s" used for service "%s" cannot be found.', $class, $id));
+            }
+
+            if (!$r->isSubclassOf(Command::class)) {
+                if (!$r->hasMethod('__invoke')) {
+                    throw new InvalidArgumentException(\sprintf('The service "%s" tagged "%s" must either be a subclass of "%s" or have an "__invoke()" method.', $id, 'console.command', Command::class));
                 }
-                $aliases = str_replace('%', '%%', $class::getDefaultName() ?? '');
+
+                $invokableRef = new Reference($id);
+                $definition = $container->register($id .= '.command', $class = Command::class)
```

Comment:
rather than making the invokable command a new (internal) primitive used by `Command`, wouldn't it make more sense to have a factory to create the desired `Command`? This way you keep `Command` as the primary primitive and it makes it easier to extend this kind of DX feature later on. For instance to support a different invoke method than `__invoke()` or add support for `interactive/initialize`